### PR TITLE
[4.1] Making Google happy about breadcrumbs again.

### PR DESCRIPTION
**Pull request by crystalenka**
This PR fixes structured data errors that made Google flag breadcrumbs as invalid.

When a breadcrumb item was a menu heading, instead of a link, Google would flag it as an invalid breadcrumb because it did not have a link. This is fixed by removing the JSON markup for items without a link, with the exception of the last item/current page.

This pull request fixes issues #37508 and #37510.

Basically reproducing my changes from my prior 4.0 pull request, but updated for the new JSON format.

### Summary of Changes

1. Added a check if a link exists in a breadcrumb item. If not, skip it and don't add it to the JSON.
2. Added the last item (aka current page) to the JSON without a link. This is valid per [Google SEO guidelines](https://developers.google.com/search/docs/advanced/structured-data/breadcrumb#list-item); they will pull the current page URL for the last item.

### Testing Instructions

Create a site menu with a menu heading and subpages, like so:

- Home
- Menu Heading
  - Subpage 1

Create a breadcrumbs module. Ensure the module is published and visible on Subpage 1.

Navigate to subpage 1.

If this site is accessible over the internet:

Copy the URL and paste it into the field here: https://search.google.com/test/rich-results

If this site is hosted locally:

View source of Subpage 1 and copy it into the code box on https://search.google.com/test/rich-results
(Alternatively, you can find the generated JSON for the module and just copy that into the code box.)

### Actual result BEFORE applying this Pull Request

Google detects the breadcrumbs but shows an error complaining about the missing URL in an item.

### Expected result AFTER applying this Pull Request

Google correctly detects the breadcrumbs and marks them as valid.

### Documentation Changes Required

None as far as I am aware.

### Additional Notes

I intentionally am not checking if the module is set to display the last item or not because for the purposes of the JSON, it is better to include it and does not affect the front end display of the module.

### CODE REVIEW

File: modules/mod_breadcrumbs/tmpl/default.php [hunk 1]
```diff
@@ -78,14 +78,27 @@
...
-		$data['itemListElement'][] = [
-				'@type'    => 'ListItem',
-				'position' => $key + 1,
-				'item'     => [
-						'@id'  => $item->link ? Route::_($item->link, true, Route::TLS_IGNORE, true) : Route::_(Uri::getInstance()),
-						'name' => $item->name
-				]
-		];
+		// Only add item to JSON if it has a valid link, otherwise skip it.
+		if(!empty($item->link)) {
```
Comment (richard67) [inline modules/mod_breadcrumbs/tmpl/default.php]:
```suggestion
		if (!empty($item->link))
		{
```
The code style checker skips layout files, but in the pure PHP parts we still should respect our code style.

File: modules/mod_breadcrumbs/tmpl/default.php [hunk 2]
```diff
@@ -78,14 +78,27 @@
...
+			$data['itemListElement'][] = [
+					'@type'    => 'ListItem',
+					'position' => $key + 1,
+					'item'     => [
+							'@id'  => $item->link ? Route::_($item->link, true, Route::TLS_IGNORE, true) : Route::_(Uri::getInstance()),
+							'name' => $item->name
+					]
+			];
+		} elseif ($key === $last_item_key) {
```
Comment (richard67) [inline modules/mod_breadcrumbs/tmpl/default.php]:
```suggestion
		}
		elseif ($key === $last_item_key)
		{
```
The code style checker skips layout files, but in the pure PHP parts we still should respect our code style.

File: modules/mod_breadcrumbs/tmpl/default.php [hunk 3]
```diff
@@ -78,14 +78,27 @@
...
-		];
+		// Only add item to JSON if it has a valid link, otherwise skip it.
+		if(!empty($item->link)) {
+			$data['itemListElement'][] = [
+					'@type'    => 'ListItem',
+					'position' => $key + 1,
+					'item'     => [
+							'@id'  => $item->link ? Route::_($item->link, true, Route::TLS_IGNORE, true) : Route::_(Uri::getInstance()),
+							'name' => $item->name
+					]
```
Comment (richard67) [inline modules/mod_breadcrumbs/tmpl/default.php]:
```suggestion
							'name' => $item->name,
					],
```
The code style checker skips layout files, but in the pure PHP parts we still should respect our code style.

**Review by richard67 — CHANGES_REQUESTED** [review]

File: modules/mod_breadcrumbs/tmpl/default.php [hunk 4]
```diff
@@ -78,14 +78,30 @@
...
+		elseif ($key === $last_item_key)
+		{
+			// Add the last item (current page) to JSON, but without a link.
+			// Google accepts items without a URL only as the current page.
+			$data['itemListElement'][] = [
+					'@type'    => 'ListItem',
+					'position' => $key + 1,
+					'item'     => [
+							'name' => $item->name
+					]
```
Comment (richard67) [inline modules/mod_breadcrumbs/tmpl/default.php]:
```suggestion
							'name' => $item->name,
					],
```
Sorry, this one I have forgotten in my last review. Of course also here for the last array elements a comma at the end due to our code style rules.

File: modules/mod_breadcrumbs/tmpl/default.php [hunk 5]
```diff
@@ -78,14 +78,30 @@
...
-				]
-		];
+		// Only add item to JSON if it has a valid link, otherwise skip it.
+		if (!empty($item->link))
+		{
+			$data['itemListElement'][] = [
+					'@type'    => 'ListItem',
+					'position' => $key + 1,
+					'item'     => [
+							'@id'  => $item->link ? Route::_($item->link, true, Route::TLS_IGNORE, true) : Route::_(Uri::getInstance()),
```
Comment (richard67) [inline modules/mod_breadcrumbs/tmpl/default.php]: Does it still need the ternary operator here, now as with the if condition in line 82 `$item->link` cannot be empty at this place?

In line 56 where we have the same situation there is no ternary.

**Comment by nadjak77** [comment]
I have tested this item :red_circle: unsuccessfully on f034a3c4c9f705698873af506d25525d1bc00f7f<br /><br />I still get the error "Field itemListElement missing"<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

**Comment by crystalenka** [comment]
Hmmm I will take a look again, maybe I missed something.

**Comment by crystalenka** [comment]
> I have tested this item 🔴 unsuccessfully on [f034a3c](https://github.com/joomla/joomla-cms/commit/f034a3c4c9f705698873af506d25525d1bc00f7f)I still get the error "Field itemListElement missing"

I just tested on a clean install of 4.1 with Cassiopeia template and the blog sample data. No errors for me: https://search.google.com/test/rich-results/result?id=JGXsLee4zFK1Py_gDrovhg 

@nadjak77 Can you please share your rich test results link with me by hitting the "share" button so I can debug? 

Screen Shot 2022-05-02 at 17 02 41

I'm not sure how you could have gotten different results, unless maybe you have a template override on the breadcrumbs?

Thanks!

File: modules/mod_breadcrumbs/tmpl/default.php (hunk 5 above)
Comment (crystalenka) [reply modules/mod_breadcrumbs/tmpl/default.php]: Good point! I'll revise.

**Review by richard67 — APPROVED** [review]

**Comment by nikosdion** [comment]
I have tested this item :red_circle: unsuccessfully on c69c9330225f1a075696b5d09ee1bdff8ada351d<br /><br />I also get the “Missing field "itemListElement"” error at https://search.google.com/test/rich-results/result/r%2Fbreadcrumbs?id=iPMy7EevFSOnJtC0n2iV5g for my test page https://nikosdiontest.joomla.com/index.php/sample-layouts/blog

However, this appears to come not from the LD-JSON microdata but from the leftover Schema.org microdata on the breadcrumbs' `<ol>` tag:

```html
<ol itemscope itemtype="https://schema.org/BreadcrumbList" class="mod-breadcrumbs breadcrumb px-3 py-2">
```

Since we're adding LD-JSON shouldn't we remove the Schema.org microdata altogether?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

**Comment by crystalenka** [comment]
> Since we're adding LD-JSON shouldn't we remove the Schema.org microdata altogether?

Yup. I missed that, must be remnant from the original conversion to JSON instead of micro data.

**Comment by nikosdion** [comment]
I have tested this item :white_check_mark: successfully on f0f4c764b51bc9848f93164a39d4f8d30faa717f<br /><br />I can confirm that now it works as described. There are no more breadcrumb microdata errors for the same page I tested before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

**Comment by richard67** [comment]
@nadjak77 Could you test again with the latest changes? Thanks in advance.

**Comment by richard67** [comment]
@aint-broke Please test if this PR solves your issue #37510 . Thanks in advance.

**Comment by nadjak77** [comment]
I have tested this item :white_check_mark: successfully on f0f4c764b51bc9848f93164a39d4f8d30faa717f<br /><br />tested successfully. Thank you<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

**Comment by nadjak77** [comment]
@crystalenka Sorry for my late reply.
My first test (unsuccessfully) was on a clean new website without any other installations. The installation was on localhost.
But now the test was successfull. Thank you.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37679">issues.joomla.org/tracker/joomla-cms/37679</a>.</sub>

**Comment by crystalenka** [comment]
No problem, @nadjak77 !! We got it figured out!

**Review by richard67 — APPROVED** [review]

**Comment by laoneo** [comment]
Thanks!

**Pull request merged**